src/compiler/parser.ts

```diff
@@ -2247,6 +2247,13 @@ namespace ts {
                 property.name = name;
                 property.questionToken = questionToken;
                 property.type = parseTypeAnnotation();
+
+                // Although object type properties cannot not have initializers, we attempt to parse an initializer
+                // so we can report that an object type property cannot have an initializer.
+                if (token === SyntaxKind.EqualsToken && lookAhead(() => parseNonParameterInitializer()) !== undefined) {

```

Comment:
I would not do it this way.  I would just parse out an initialize for these properties.  Then, in the checker, detect that and report a nice error there.

We do not want random errors being produced the parser like this.